File: app/src/main/java/com/sumian/sd/tab/MeFragment.java
package com.sumian.sd.tab;

import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModelProviders;
import android.text.TextUtils;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.blankj.utilcode.util.ActivityUtils;
import com.hyphenate.helpdesk.easeui.UIProvider;
import com.sumian.common.image.ImageLoader;
import com.sumian.hw.device.bean.BlueDevice;
import com.sumian.hw.leancloud.HwLeanCloudHelper;
import com.sumian.hw.upgrade.model.VersionModel;
import com.sumian.sd.R;
import com.sumian.sd.account.bean.UserInfo;
import com.sumian.sd.account.userProfile.SdUserProfileActivity;
import com.sumian.sd.app.App;
import com.sumian.sd.app.AppManager;
import com.sumian.sd.base.SdBaseFragment;
import com.sumian.sd.device.DeviceManageActivity;
import com.sumian.sd.device.DeviceManager;
import com.sumian.sd.h5.SleepFileWebActivity;
import com.sumian.sd.kefu.KefuManager;
import com.sumian.sd.main.OnEnterListener;
import com.sumian.sd.notification.NotificationListActivity;
import com.sumian.sd.notification.NotificationViewModel;
import com.sumian.sd.onlinereport.OnlineReportListActivity;
import com.sumian.sd.scale.ScaleListActivity;
import com.sumian.sd.service.advisory.activity.AdvisoryListActivity;
import com.sumian.sd.service.diary.DiaryEvaluationListActivity;
import com.sumian.sd.service.tel.activity.TelBookingListActivity;
import com.sumian.sd.setting.SettingActivity;
import com.sumian.sd.theme.three.SkinConfig;
import com.sumian.sd.widget.divider.SettingDividerView;
import com.sumian.sd.widget.tips.PatientRecordTips;
import com.sumian.sd.widget.tips.PatientServiceTips;

import org.jetbrains.annotations.Nullable;

import java.util.Objects;

import butterknife.BindView;
import butterknife.OnClick;
import de.hdodenhof.circleimageview.CircleImageView;

/**
 * Created by jzz
 * on 2018/1/16.
 * desc:
 */

@SuppressWarnings("ALL")
public class MeFragment extends SdBaseFragment implements View.OnClickListener, PatientServiceTips.OnServiceTipsCallback,
        PatientRecordTips.OnRecordTipsCallback, HwLeanCloudHelper.OnShowMsgDotCallback, OnEnterListener, VersionModel.ShowDotCallback {

    @BindView(R.id.iv_avatar)
    CircleImageView mIvAvatar;
    @BindView(R.id.tv_nickname)
    TextView mTvNickname;
    @BindView(R.id.iv_notification)
    ImageView mIvNotification;

    //@BindView(R.id.siv_customer_service)
    //ImageView mSivKefu;

    @BindView(R.id.tips_service)
    PatientServiceTips mTipsService;
    @BindView(R.id.tips_record)
    PatientRecordTips mTipsRecord;
    @BindView(R.id.dv_device_manage)
    SettingDividerView mSdDeviceManage;

    @Override
    protected int getLayoutId() {
        return R.layout.fragment_tab_me;
    }

    @Override
    protected void initWidget(View root) {
        super.initWidget(root);
        this.mTipsService.setOnServiceTipsCallback(this);
        this.mTipsRecord.setOnRecordTipsCallback(this);
    }

    @Override
    protected void initData() {
        super.initData();
        UserInfo userProfile = AppManager.getAccountViewModel().getToken().user;
        updateUserProfile(userProfile);
        AppManager.getAccountViewModel().getLiveDataToken().observe(this, token -> {
            if (token != null) {
                updateUserProfile(token.user);
            }
        });
        ViewModelProviders.of(Objects.requireNonNull(getActivity()))
                .get(NotificationViewModel.class)
                .getUnreadCount()
                .observe(this, unreadCount -> mIvNotification.setActivated(unreadCount != null && unreadCount > 0));

        // HwLeanCloudHelper.addOnAdminMsgCallback(this);
        DeviceManager.INSTANCE.getMonitorLiveData().observe(this, new Observer<BlueDevice>() {
            @Override
            public void onChanged(@android.support.annotation.Nullable BlueDevice blueDevice) {
                String monitorSn = blueDevice == null ? null : blueDevice.sn;
                if (TextUtils.isEmpty(monitorSn)) {
                    monitorSn = getString(R.string.add_new_device);
                }
                mSdDeviceManage.setContent(monitorSn);
            }
        });

    }

    @Override
    @OnClick({R.id.iv_avatar, R.id.tv_nickname, R.id.dv_setting, R.id.iv_notification, R.id.siv_customer_service, R.id.dv_device_manage})
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.iv_avatar:
            case R.id.tv_nickname:
                SdUserProfileActivity.show(getContext(), SdUserProfileActivity.class);
                break;
            case R.id.dv_setting:
                SettingActivity.show(getContext(), SettingActivity.class);
                break;
            case R.id.iv_notification:
                NotificationListActivity.launch(getActivity());
                break;
            case R.id.siv_customer_service:
                UIProvider.getInstance().clearCacheMsg();
                UIProvider.getInstance().setThemeMode(SkinConfig.isInNightMode(App.getAppContext()) ? 0x02 : 0x01);
                KefuManager.launchKefuActivity();
                break;
            case R.id.dv_device_manage:
                ActivityUtils.startActivity(DeviceManageActivity.class);
                break;
            default:
                break;
        }
    }

    private void updateUserProfile(UserInfo userProfile) {
        ImageLoader.loadImage(userProfile.avatar, mIvAvatar, R.mipmap.ic_info_avatar_patient);
        String nickname = userProfile.nickname;
        mTvNickname.setText(TextUtils.isEmpty(nickname) ? userProfile.mobile : nickname);
    }

    @Override
    public void showGraphicService() {
        AdvisoryListActivity.show(getActivity(), AdvisoryListActivity.class);
    }

    @Override
    public void showTelService() {
        TelBookingListActivity.show();
    }

    @Override
    public void onDiaryEvaluationClick() {
        ActivityUtils.startActivity(DiaryEvaluationListActivity.class);
    }

    @Override
    public void showSleepRecord() {
        SleepFileWebActivity.show(getContext());
    }

    @Override
    public void showEvaluation() {
        ScaleListActivity.launch(getContext(), ScaleListActivity.TYPE_FILLED);
    }

    @Override
    public void showOnlineReport() {
        OnlineReportListActivity.launchForShowAll(this);
    }

    @Override
    public void onShowMsgDotCallback(int adminMsgLen, int doctorMsgLen, int customerMsgLen) {
        // onHideMsgCallback(adminMsgLen, doctorMsgLen, customerMsgLen);
    }

    @Override
    public void onHideMsgCallback(int adminMsgLen, int doctorMsgLen, int customerMsgLen) {
        // runOnUiThread(() -> mSivKefu.setImageResource((customerMsgLen > 0) ? R.drawable.ic_info_customerservice_reply : R.drawable.ic_info_customerservice));
    }

    @Override
    public void onEnter(@Nullable String data) {
        // if (mSivKefu != null) {
        //    runOnUiThread(() -> mSivKefu.setImageResource((HwLeanCloudHelper.isHaveCustomerMsg()) ? R.drawable.ic_info_customerservice_reply : R.drawable.ic_info_customerservice));
        // }
    }

    @Override
    public void showDot(boolean isShowAppDot, boolean isShowMonitorDot, boolean isShowSleepyDot) {

    }
}

File: app/src/main/java/com/sumian/sd/tab/MeFragment.kt
package com.sumian.sd.tab

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.support.v4.app.FragmentActivity
import android.text.TextUtils
import android.view.View
import com.blankj.utilcode.util.ActivityUtils
import com.hyphenate.helpdesk.easeui.UIProvider
import com.sumian.common.image.ImageLoader
import com.sumian.hw.leancloud.HwLeanCloudHelper
import com.sumian.hw.log.LogManager
import com.sumian.hw.upgrade.model.VersionModel
import com.sumian.sd.R
import com.sumian.sd.account.bean.Token
import com.sumian.sd.account.bean.UserInfo
import com.sumian.sd.account.userProfile.SdUserProfileActivity
import com.sumian.sd.app.AppManager
import com.sumian.sd.base.SdBaseFragment
import com.sumian.sd.base.SdBasePresenter
import com.sumian.sd.device.DeviceManageActivity
import com.sumian.sd.device.DeviceManager
import com.sumian.sd.h5.SleepFileWebActivity
import com.sumian.sd.kefu.KefuManager
import com.sumian.sd.main.OnEnterListener
import com.sumian.sd.notification.NotificationListActivity
import com.sumian.sd.notification.NotificationViewModel
import com.sumian.sd.onlinereport.OnlineReportListActivity
import com.sumian.sd.scale.ScaleListActivity
import com.sumian.sd.service.advisory.activity.AdvisoryListActivity
import com.sumian.sd.service.coupon.activity.CouponCenterActivity
import com.sumian.sd.service.diary.DiaryEvaluationListActivity
import com.sumian.sd.service.tel.activity.TelBookingListActivity
import com.sumian.sd.setting.SettingActivity
import com.sumian.sd.widget.tips.PatientRecordTips
import com.sumian.sd.widget.tips.PatientServiceTips
import kotlinx.android.synthetic.main.fragment_tab_me.*
import java.util.*

/**
 * Created by jzz
 * on 2018/1/16.
 * desc:
 */

class MeFragment : SdBaseFragment<SdBasePresenter<*>>(), View.OnClickListener, PatientServiceTips.OnServiceTipsCallback, PatientRecordTips.OnRecordTipsCallback,
        HwLeanCloudHelper.OnShowMsgDotCallback, OnEnterListener, VersionModel.ShowDotCallback {

    //@BindView(R.id.siv_customer_service)
    //ImageView mSivKefu;

    override fun getLayoutId(): Int {
        return R.layout.fragment_tab_me
    }

    override fun initWidget(root: View) {
        super.initWidget(root)
        iv_avatar.setOnClickListener(this)
        tv_nickname.setOnClickListener(this)
        dv_setting.setOnClickListener(this)
        iv_notification.setOnClickListener(this)
        siv_customer_service.setOnClickListener(this)
        dv_device_manage.setOnClickListener(this)
        dv_coupon_center.setOnClickListener(this)
        tips_service.setOnServiceTipsCallback(this)
        tips_record.setOnRecordTipsCallback(this)
    }

    override fun initData() {
        super.initData()
        val userProfile = AppManager.getAccountViewModel().token.user
        updateUserProfile(userProfile)
        AppManager.getAccountViewModel().liveDataToken.observe(this, Observer<Token> { token ->
            run {
                token?.let {
                    updateUserProfile(it.user)
                }
            }
        })
        ViewModelProviders.of(Objects.requireNonNull<FragmentActivity>(activity))
                .get(NotificationViewModel::class.java)
                .unreadCount
                .observe(this, Observer<Int> { unreadCount -> iv_notification.isActivated = unreadCount != null && unreadCount > 0 })

        // HwLeanCloudHelper.addOnAdminMsgCallback(this);
        DeviceManager.getMonitorLiveData().observe(this, Observer { blueDevice ->
            var monitorSn: String? = blueDevice?.sn
            if (TextUtils.isEmpty(monitorSn)) {
                monitorSn = getString(R.string.add_new_device)
            }
            dv_device_manage.setContent(monitorSn!!)
        })

    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.iv_avatar, R.id.tv_nickname -> SdUserProfileActivity.show(context, SdUserProfileActivity::class.java)
            R.id.dv_setting -> SettingActivity.show(context, SettingActivity::class.java)
            R.id.iv_notification -> NotificationListActivity.launch(activity)
            R.id.siv_customer_service -> {
                UIProvider.getInstance().clearCacheMsg()
                KefuManager.launchKefuActivity()
            }
            R.id.dv_device_manage -> {
                LogManager.appendUserOperationLog("点击【设备管理】")
                ActivityUtils.startActivity(DeviceManageActivity::class.java)
            }
            R.id.dv_coupon_center -> CouponCenterActivity.show()
            else -> {
            }
        }
    }

    private fun updateUserProfile(userProfile: UserInfo) {
        ImageLoader.loadImage(userProfile.avatar, iv_avatar, R.mipmap.ic_info_avatar_patient)
        val nickname = userProfile.nickname
        tv_nickname.text = if (TextUtils.isEmpty(nickname)) userProfile.mobile else nickname
    }

    override fun showGraphicService() {
        AdvisoryListActivity.show()
    }

    override fun showTelService() {
        TelBookingListActivity.show()
    }

    override fun onDiaryEvaluationClick() {
        ActivityUtils.startActivity(DiaryEvaluationListActivity::class.java)
    }

    override fun showSleepRecord() {
        SleepFileWebActivity.show(context!!)
    }

    override fun showEvaluation() {
        ScaleListActivity.launch(1)
    }

    override fun showOnlineReport() {
        OnlineReportListActivity.launchForShowAll(this)
    }

    override fun onShowMsgDotCallback(adminMsgLen: Int, doctorMsgLen: Int, customerMsgLen: Int) {
        // onHideMsgCallback(adminMsgLen, doctorMsgLen, customerMsgLen);
    }

    override fun onHideMsgCallback(adminMsgLen: Int, doctorMsgLen: Int, customerMsgLen: Int) {
        // runOnUiThread(() -> mSivKefu.setImageResource((customerMsgLen > 0) ? R.drawable.ic_info_customerservice_reply : R.drawable.ic_info_customerservice));
    }

    override fun onEnter(data: String?) {
        // if (mSivKefu != null) {
        //    runOnUiThread(() -> mSivKefu.setImageResource((HwLeanCloudHelper.isHaveCustomerMsg()) ? R.drawable.ic_info_customerservice_reply : R.drawable.ic_info_customerservice));
        // }
    }

    override fun showDot(isShowAppDot: Boolean, isShowMonitorDot: Boolean, isShowSleepyDot: Boolean) {

    }
}
